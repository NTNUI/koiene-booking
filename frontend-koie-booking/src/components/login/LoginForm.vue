<template>
  <div :class="$style.wrapper">
    <form style="display:flex; flex-direction: column;" @submit.prevent="login">
      <label :class="$style.label__description" for="username">
        {{ $t('ntnui.login.email') }}
        <form-input
          id="username"
          :class="$style.username__field"
          :input.sync="email"
          :valid="isValidEmail"
          autocomplete="username"
        />
      </label>
      <label :class="$style.label__description" for="password">
        {{ $t('ntnui.login.password') }}
        <form-input
          id="password"
          input-type="password"
          :input.sync="password"
          :valid="isValidPassword"
          autocomplete="current-password"
        />
      </label>
      <div v-show="error" :class="$style.error">
        <p>{{ error }}</p>
      </div>
      <v-btn type="submit" :color="this.$scssVars.globalColorPrimary" :loading="loading">
        {{ $t('ntnui.login.login') }}
      </v-btn>
    </form>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import Input from '@/components/login/Input.vue';
import { fetchToken } from '@/service/auth';
import { TokenPayload } from '@/types/tokenPayload';
import { loginFormData } from '@/types/login';

export default Vue.extend({
  name: 'LoginForm',
  components: {
    'form-input': Input,
  },
  data(): loginFormData {
    return {
      email: '',
      password: '',
      error: '',
      loading: false,
    };
  },
  computed: {
    isValidEmail(): boolean {
      return this.email.includes('@') && this.email.includes('.');
    },
    isValidPassword(): boolean {
      return this.password.length >= 8;
    },
  },
  methods: {
    async login() {
      if (!this.isValidEmail && !this.loading) {
        this.error = `${this.$t('ntnui.login.error.emailInvalid')}`;
        setTimeout(() => (this.error = ''), 5000);
      } else if (!this.isValidPassword && !this.loading) {
        this.error = `${this.$t('ntnui.login.error.length')}`;
      } else if (!this.loading) {
        const payload: TokenPayload = {
          email: this.email,
          password: this.password,
        };
        try {
          this.loading = true;
          await fetchToken(payload);
          this.$router.push(
            this.checkIfUserIsKeyManagerAndNotAdmin()
              ? '/key_management'
              : (this.$route.query.redirect as string) || '/'
          );
        } catch (error) {
          this.$notify({
            group: 'memberships_system',
            type: 'error',
            duration: 3000,
            text: `${this.$t('ntnui.login.error.noUser')}`,
          });
        } finally {
          this.loading = false;
        }
      }
    },
    checkIfUserIsKeyManagerAndNotAdmin(): boolean {
      return !this.$store.getters['auth/isAdmin'] && this.$store.getters['auth/isKeyManager'];
    },
  },
});
</script>

<style lang="scss" module>
.wrapper {
  width: 100%;
  margin-top: 30px;
}
.horizontal__ruler {
  display: block;
  unicode-bidi: isolate;
  margin-block-start: 0.5em;
  margin-block-end: 0.5em;
  margin-inline-start: auto;
  margin-inline-end: auto;
  overflow: hidden;
  border-style: inset;
  border-width: 1px;
}
.label__description {
  font-weight: 400;
}
.info__circle {
  color: yellow;
  margin-right: 15px;
  font-size: 22px;
}
.username__field {
  margin-bottom: 10px;
  border-radius: 20px;
}
.error {
  display: flex;
  justify-content: center;
  align-items: center;
  color: white;
  text-align: center;
  margin-top: 15px;
  font-size: 14px;
}
.submit__button {
  margin-top: 20px;
}
</style>
