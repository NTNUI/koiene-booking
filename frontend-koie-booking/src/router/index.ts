import Vue from 'vue';
import VueRouter from 'vue-router';
import store from '@/store';

Vue.use(VueRouter);

const routes = [
  {
    path: '/',
    name: 'AllKoier',
    component: () => import(/* webpackChunkName: "about" */ '../views/AllKoier.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/koie/:id',
    name: 'koie',
    component: () => import('../views/Koie.vue'),
    props: true,
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/booking/:id',
    name: 'booking',
    component: () => import('../views/Booking.vue'),
    meta: { requiresAuth: true, requiresAdmin: false }
  },
  {
    path: '/login',
    name: 'Login',
    component: () => import('../views/Login.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/admin',
    name: 'Admin',
    component: () => import('../views/Admin.vue'),
    meta: { requiresAuth: true, requiresAdmin: true }
  }
];

const router = new VueRouter({
  routes
});

export function beforeEach(to: any, from: any, next: any) {
  const isLoggedIn = store.getters['auth/isLoggedIn'];
  const isAdmin = store.getters['auth/isAdmin'];
  const requiresAuth = to.matched.some((record: any) => record.meta.requiresAuth);
  const requiresAdmin = to.matched.some((record: any) => record.meta.requiresAdmin);

  if (requiresAdmin && !isAdmin) next({ name: 'Login', query: { redirect: to.path } });
  else if (requiresAuth && !isLoggedIn) next({ name: 'Login', query: { redirect: to.path } });
  else if (to.name === 'Login' && isLoggedIn) next('/');
  else next();
}

router.beforeEach((to, from, next) => beforeEach(to, from, next));

export default router;
export { routes };
