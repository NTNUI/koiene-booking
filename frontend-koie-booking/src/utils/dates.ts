import dayjs from 'dayjs';
require('dayjs/locale/nb');
dayjs.locale('nb');

export function getDateString(baseDate?: string, addDays?: number): string {
  let resultString: string;
  if (baseDate && addDays) {
    resultString = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else if (baseDate) {
    resultString = new Date(baseDate).toISOString().substr(0, 10);
  } else if (addDays) {
    resultString = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else {
    resultString = new Date().toISOString().substr(0, 10);
  }
  return resultString;
}

export function getDate(baseDate?: string, addDays?: number): Date {
  let resultDate: Date;
  if (baseDate && addDays) {
    resultDate = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000);
  } else if (baseDate) {
    resultDate = new Date(baseDate);
  } else if (addDays) {
    resultDate = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000);
  } else {
    resultDate = new Date();
  }
  return resultDate;
}

export function addToDate(startDate: string, howMany: number, timeUnit: dayjs.OpUnitType): string {
  const date = dayjs(startDate + 'T00:00:00.000Z');
  return date.add(howMany, timeUnit).format('YYYY-MM-DD');
}

export function formatDate(dateISO: string, formatString: string): string {
  const date = dayjs(dateISO + 'T00:00:00.000Z');
  return date.format(formatString);
}
