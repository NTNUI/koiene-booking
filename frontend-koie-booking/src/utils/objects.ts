export const clone = (object: any) => Object.assign({}, object);

export const renameKey = (object: any, key: string, newKey: string): any => {
  const value = object[key];
  delete object[key];
  object[newKey] = value;
  return object;
};

export const sortObjectByKey = (unordered: any) => {
  let ordered: any;
  ordered = {};
  Object.keys(unordered)
    .sort()
    .forEach(function(key) {
      ordered[key] = unordered[key];
    });
  return ordered;
};
