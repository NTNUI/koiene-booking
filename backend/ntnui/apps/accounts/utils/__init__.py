from accounts.utils.mail_utils import send_gsuite_confirmation_mail
