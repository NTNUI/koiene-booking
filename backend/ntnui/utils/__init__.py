from ntnui.utils.pagination import PageSize7, PageSize20, PageSize100
