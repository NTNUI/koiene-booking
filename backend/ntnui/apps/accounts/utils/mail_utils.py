from django.core.mail import send_mail
from django.template.loader import render_to_string

from ntnui.enums import MembershipType


def new_membership(membership):
    """ Sends confirmation email to users and guests who signs up for events and sub-events. """

    # Mail header.
    sender = "noreply@mg.ntnui.no"
    receiver = [membership.member.email]
    subject = "Welcome"

    context = {"user": membership.member, "group": membership.group}

    # Mail body.
    plain_message = render_to_string("emails/new_membership.txt", context)
    html_message = render_to_string("emails/new_membership.html", context)

    send_mail(subject, plain_message, sender, receiver, html_message=html_message)


def send_gsuite_confirmation_mail(user, contact_email, password):
    # Mail header.
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = "Opprettelse av ny NTNUI-bruker"

    context = {
        "first_name": user.first_name,
        "email": user.email,
        "contact_email": contact_email,
        "password": password,
    }

    # Mail body.
    plain_message = render_to_string("emails/new_gsuite_account.txt", context)
    html_message = render_to_string("emails/new_gsuite_account.html", context)

    send_mail(subject, plain_message, sender, receiver, html_message=html_message)


def send_welcome_mail_to_leader(user, group):
    """ Send a mail to a new leader """
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny leder i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_leader_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_leader_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail_to_cashier(user, group):
    """Send a mail to the new cashier"""
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny kasserer i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_cashier_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_cashier_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail_to_board_member(user, group):
    """Send a mail to the new cashier"""
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny tillitsvalgt i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_board_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_board_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail(membership):
    """ Delegate welcome mail to different board members """
    send_welcome_mail_to_board_member(user=membership.member, group=membership.group)
    if membership.type == MembershipType.leader:
        send_welcome_mail_to_leader(user=membership.member, group=membership.group)

    if membership.type == MembershipType.cashier:
        send_welcome_mail_to_cashier(user=membership.member, group=membership.group)
