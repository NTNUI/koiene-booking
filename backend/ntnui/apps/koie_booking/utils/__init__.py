from koie_booking.utils.mail_utils import send_confirmation_mail, send_koie_information_mail
from koie_booking.utils.date_utils import get_daterange
