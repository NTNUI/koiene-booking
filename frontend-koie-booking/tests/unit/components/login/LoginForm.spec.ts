import Vue from 'vue';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';
import { routes } from '@/router';

Vue.use(Vuetify);

// Components or views
import LoginForm from '@/components/login/LoginForm.vue';
import { createWrapper } from '../../utils';

describe('Component LoginForm.vue', () => {
  const router = new VueRouter({ routes, mode: 'abstract' });

  it('Matches snapshot', async () => {
    const wrapper = createWrapper(LoginForm, {
      router,
    });

    expect(wrapper).toMatchSnapshot();
  });
});
