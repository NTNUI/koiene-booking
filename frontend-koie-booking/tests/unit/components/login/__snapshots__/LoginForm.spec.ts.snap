// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component LoginForm.vue Matches snapshot 1`] = `
<div class="wrapper">
  <form style="display: flex; flex-direction: column;"><label for="username" class="label__description">
      EMAIL
      <div id="username" class="username__field">
        <div></div>
        <div><input autocomplete="username" type="text" class="input"></div>
      </div></label> <label for="password" class="label__description">
      PASSWORD
      <div id="password">
        <div></div>
        <div><input autocomplete="current-password" type="password" class="input"></div>
      </div></label>
    <div class="error" style="display: none;">
      <p></p>
    </div> <button type="submit" class="v-btn v-btn--contained theme--light v-size--default"><span class="v-btn__content">
      LOG IN
    </span></button>
  </form>
</div>
`;
